feat(behavior_path_planner): use object recognition for pull_over

kosuke55 opened the pull request:
## Description

<!-- Write a brief description of this PR. -->
Use object recognition for pull_over
This is a solution to the problem of not being able to park on the shoulder due to vegetation.
- Both occupancy gird and object recognition can be used. Both can be optionally enabled.
  - `use_occupancy_grid`
  - `use_object_recognition`

- The visualization of the division of the goal search area was stopped to reduce the computational complexity.

https://user-images.githubusercontent.com/39142679/188359594-c6724e3e-1cb7-4051-9a18-8d2c67d4dee9.mp4

## Related links

<!-- Write the links related to this PR. -->

## Tests performed
- psim 
- lsim (TIER IV internal rosbag) 

The original goal is also repositioned.

https://console.data-search.tier4.jp/projects/prd_jt/environments/9ba5b2bc-9b8e-438a-aba6-342d19c876a5/rosbags/7d31ef3c-5cae-4728-9403-975f911c7ee9

https://user-images.githubusercontent.com/39142679/188359739-42925a38-29b8-4b2c-9e2c-457021cab5b8.mp4

https://console.data-search.tier4.jp/projects/prd_jt/environments/9ba5b2bc-9b8e-438a-aba6-342d19c876a5/rosbags/85b8a4ee-f2b9-4312-814c-786e98b5ce7e

https://user-images.githubusercontent.com/39142679/188359782-2b6a52fd-235f-4d99-89cf-12f99f622bce.mp4

<!-- Describe how you have tested this PR. -->

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

shmpwk commented on planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:
```diff
@@ -341,6 +347,49 @@ bool PullOverModule::isLongEnoughToParkingStart(
+      return true;
+    }
+  }
+
+  // no collsion
```
[minor suggestion]

```suggestion
```

shmpwk commented on planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp:
```diff
@@ -209,6 +215,8 @@ class PullOverModule : public SceneModuleInterface
   void updateOccupancyGrid();
   void researchGoal();
   void resetStatus();
+  bool checkCollision(const Pose & pose) const;
+  bool checkCollision(const PathWithLaneId & path) const;
```
I am not sure, but if it is a template function, I understand that they are the same name.
But if not, I was wondering if the names should not be the same.

kosuke55 replied:
It's just overload. changed functions name in https://github.com/autowarefoundation/autoware.universe/pull/1777/commits/ad499414df941db5dd86900a164f3613df76b7b9

shmpwk replied:
Sorry... simple overload is not problem at all according to the google style guide :bow: 
https://google.github.io/styleguide/cppguide.html#Operator_Overloading

So the previous one is enough but the new one is of cause perfect for me.

kosuke55 commented:
@shmpwk 
Thanks for your review.
Updated docs and removed unnecessary processing. It does not affect the behavior.
https://github.com/autowarefoundation/autoware.universe/pull/1777/commits/ac7b8826a79dfe692a8c597884cf9cede3b76000
https://github.com/autowarefoundation/autoware.universe/pull/1777/commits/d2c3016815fb633918d50c9eeeab8bbef125e923

kosuke55 commented:
@shmpwk 
Updated warning message and throw exception when search_priority is invalid.
https://github.com/autowarefoundation/autoware.universe/pull/1777/commits/1dcd4c3f3f75c609373009c57cb7251ed4bcf579
